### PR TITLE
AIP 6: Security council Elections

### PR DESCRIPTION
**Pull request by DZGoldman**
implementation of Security Council elections 
https://forum.arbitrum.foundation/t/aip-6-security-council-elections/15425

this includes fixes from audit reports produced by Trail of Bits and Code4Rena

### CODE REVIEW

File: src/security-council-mgmt/SecurityCouncilManager.sol [hunk 1]
```diff
@@ -0,0 +1,403 @@
...
+    /// @inheritdoc ISecurityCouncilManager
+    function getBothCohorts() public view returns (address[] memory) {
+        address[] memory members = new address[](firstCohort.length + secondCohort.length);
+        for (uint256 i = 0; i < firstCohort.length; i++) {
+            members[i] = firstCohort[i];
+        }
+        for (uint256 i = 0; i < secondCohort.length; i++) {
+            members[firstCohort.length + i] = secondCohort[i];
+        }
+        return members;
```
Comment (gzeoneth) [inline src/security-council-mgmt/SecurityCouncilManager.sol:330]:
```suggestion
        uint256 firstCohortLength = firstCohort.length;
        uint256 secondCohortLength = secondCohort.length;
        address[] memory members = new address[](firstCohortLength + secondCohortLength);
        for (uint256 i = 0; i < firstCohortLength; i++) {
            members[i] = firstCohort[i];
        }
        for (uint256 i = 0; i < secondCohortLength; i++) {
            members[firstCohortLength + i] = secondCohort[i];
        }
        return members;
```

File: src/security-council-mgmt/SecurityCouncilManager.sol [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+                SecurityCouncilUpgradeAction.perform.selector,
+                securityCouncilData.securityCouncil,
+                newMembers
+            );
+        }
+
+        (address to, bytes memory data) = router.createActionRouteData(
+            chainIds,
+            actionAddresses,
+            new uint256[](securityCouncils.length), // all values are always 0
```
Comment (gzeoneth) [inline src/security-council-mgmt/SecurityCouncilManager.sol:388]:
```suggestion
        uint256 securityCouncilsLength = securityCouncils.length;

        // build batch call to L1 timelock
        address[] memory actionAddresses = new address[](securityCouncilsLength);
        bytes[] memory actionDatas = new bytes[](securityCouncilsLength);
        uint256[] memory chainIds = new uint256[](securityCouncilsLength);

        for (uint256 i = 0; i < securityCouncilsLength; i++) {
            SecurityCouncilData memory securityCouncilData = securityCouncils[i];

            actionAddresses[i] = securityCouncilData.updateAction;
            chainIds[i] = securityCouncilData.chainId;
            actionDatas[i] = abi.encodeWithSelector(
                SecurityCouncilUpgradeAction.perform.selector,
                securityCouncilData.securityCouncil,
                newMembers
            );
        }

        (address to, bytes memory data) = router.createActionRouteData(
            chainIds,
            actionAddresses,
            new uint256[](securityCouncilsLength), // all values are always 0
```

File: src/security-council-mgmt/SecurityCouncilManager.sol [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+                revert SecurityCouncilAlreadyInRouter(_securityCouncilData);
+            }
+        }
+
+        securityCouncils.push(_securityCouncilData);
+        emit SecurityCouncilAdded(
+            _securityCouncilData.securityCouncil,
+            _securityCouncilData.updateAction,
+            securityCouncils.length
+        );
```
Comment (gzeoneth) [inline src/security-council-mgmt/SecurityCouncilManager.sol:252]:
```suggestion
        uint256 securityCouncilsLength = securityCouncils.length;
        if (securityCouncils.length == MAX_SECURITY_COUNCILS) {
            revert MaxSecurityCouncils(securityCouncils.length);
        }

        if (
            _securityCouncilData.updateAction == address(0)
                || _securityCouncilData.securityCouncil == address(0)
        ) {
            revert ZeroAddress();
        }

        if (_securityCouncilData.chainId == 0) {
            revert SecurityCouncilZeroChainID(_securityCouncilData);
        }

        if (!router.upExecLocationExists(_securityCouncilData.chainId)) {
            revert SecurityCouncilNotInRouter(_securityCouncilData);
        }

        for (uint256 i = 0; i < securityCouncilsLength; i++) {
            SecurityCouncilData storage existantSecurityCouncil = securityCouncils[i];

            if (
                existantSecurityCouncil.chainId == _securityCouncilData.chainId
                    && existantSecurityCouncil.securityCouncil == _securityCouncilData.securityCouncil
            ) {
                revert SecurityCouncilAlreadyInRouter(_securityCouncilData);
            }
        }

        securityCouncils.push(_securityCouncilData);
        emit SecurityCouncilAdded(
            _securityCouncilData.securityCouncil,
            _securityCouncilData.updateAction,
            securityCouncilsLength + 1
        );
```

File: src/UpgradeExecRouterBuilder.sol [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+    // Default args for creating a proposal, used by createProposalWithDefaulArgs and createProposalBatchWithDefaultArgs
+    bytes public constant DEFAULT_GOV_ACTION_CALLDATA =
+        abi.encodeWithSelector(DefaultGovAction.perform.selector);
+    uint256 public constant DEFAULT_VALUE = 0;
+    bytes32 public constant DEFAULT_PREDECESSOR = bytes32(0);
+
+    // address of l1 timelock for core governance route
+    address public l1TimelockAddr;
+    // min delay for core l1 timelock, to be kept equivalent to value on L1. I.e., if L1 delay is increased, this should be redeployed with the new value
+    uint256 public l1TimelockMinDelay;
```
Comment (gzeoneth) [inline src/UpgradeExecRouterBuilder.sol:48]:
```suggestion
    address public immutable l1TimelockAddr;
    // min delay for core l1 timelock, to be kept equivalent to value on L1. I.e., if L1 delay is increased, this should be redeployed with the new value
    uint256 public immutable l1TimelockMinDelay;
```

**Review by gzeoneth — COMMENTED** [review]
some small suggestions

**Review by yahgwai — APPROVED** [review]

**Review by godzillaba — APPROVED** [review]

**Review by fredlacs — APPROVED** [review]
LGTM

**Pull request merged**